### PR TITLE
chore(#183): introduce functions for easy mocking GH requests in tests

### PR DESCRIPTION
**Pull request by MatousJobanek**
In this PR I proposing a set of functions and structs that should make the setup of the gock mocks much easier and uncluttered.
You don't have to review it completely in detail, for now. I just want to open a discussion about the overall idea, API, and design. 
* there are two entry points
  * `MockPr` that takes a JSON content (can be loaded from default location or struct) and creates an instance of the `MockPrBuilder`. This is then responsible for collecting all information that is necessary for the mock creation
  * `NewMockPluginTemplate` that takes a plugins name and then provides the same method `MockPr` as mentioned above. This is for test cases when it is necessary to know which plugin is used - eg. for mocking status calls

* `MockPrBuilder` is the main building block of the whole API - it offers a set of methods - eg:
   * `WithTitle`
   * `WithFiles`
   * `WithoutFiles`
   * `WithComments`
   * `WithoutComments`
   * `WithReviews`
   * `WithLabels`
   * `WithUsers`
   * ...etc...
that sets these information to the builder and when the method `Create` is called, then based on the provided data it sets the necessary gock mocks. 
* for checking `POST` requests `MockPrBuilder` offers method `Expecting` where you can define the expected post payload for
  * `Comment`
  * `Status`
  * `RemovedLabel`
  * `AddedLabel`
* The method `Create` returns an instance of `PrMock` that offers additional methods such as:
  * `CreateCommentEvent`
  * `CreatePullRequestEvent`
  * `CreateUserPermissionService`
  this can be used for calling the plugin's logic.
The overall usage can look like:
```go
// given
prMock := mocker.MockPr(LoadedFromDefaultJSON()).
	WithFiles(LoadedFrom("test_fixtures/github_calls/prs/with_tests/changes_go_files.json")).
	WithoutRawFiles(ghservice.ConfigHome+"test-keeper_hint.md").
	WithConfigFile(
		ConfigYml(LoadedFrom("test_fixtures/github_calls/prs/with_tests/test-keeper.yml"))).
	WithoutComments().
	Expecting(
		Status(ToBe(github.StatusFailure, testkeeper.NoTestsMessage, testkeeper.NoTestsDetailsPageName)),
		Comment(ToHaveBodyWithWholePluginsComment)).
	Create()

// when
err := handler.HandleEvent(log, github.PullRequest, prMock.CreatePullRequestEvent("opened"))
```

* there is also a possibility to not use the whole `MockPrBuilder` but only part of it (with the usage of `SubMockBuilder`). An example is a usage of mocks for config files:
```go
mocker.MockConfig(ConfigYaml("test_patterns: ['*my', 'test.go', 'pattern.js']\n" +
	"skip_validation_for: ['pom.xml', 'regex{{*\\.adoc}}']\n" +
	"plugin_hint: 'http://my.server.com/message.md'")).
	ForChange(change)
```

@dipak-pawar @hemanik @bartoszmajsak WDYT about it? Is it something that makes mocking easier or not? So far, I used it for the `event_handler_gh_test` cases, for the configuration file mocking and permission requests. Do you think that we should provide more function (methods for `SubMockBuilder`) also for the "smaller" test cases? 

Fixes: #183

### CODE REVIEW

**Comment by hemanik** [comment]
I really like the idea of abstracting out the details and having a more intuitive DSL for the purpose of mocking the REST calls and enabling reuse.

My views and few minor suggestions on the above proposed DSL that I can think of are as follows: 
1.  How about having `LoadedFromDefaultJSON()` and `LoadedFromDefaultStruct()` as a separate endpoint? 
2. Using the `LoadedFrom()` inside of the `WithFiles()` or `ConfigYml()` so we could directly pass the file path.

```java
prMock := mocker.MockPR().
	         FromJSON().
		      WithFiles("test_fixtures/github_calls/prs/with_tests/changes_go_files.json").
		      WithoutRawFiles(ghservice.ConfigHome+"test-keeper_hint.md").
                      WithConfigFile(ConfigYml("test_fixtures/github_calls/prs/with_tests/test-keeper.yml")).
		      WithoutComments().
		 Expecting(
		      Status(ToBe(github.StatusFailure, testkeeper.NoTestsMessage, testkeeper.NoTestsDetailsPageName)),
                      Comment(ToHaveBodyWithWholePluginsComment)).
		 Create()
```
Also, I see that we are using the `Expecting()` part for checking matcher for the `POST`, `DELETE` or `PATCH` requests. Although it abstracts these details, but I somehow miss details  about the change like if something is deleted or updated. 
Rethinking on it with the DSL design may be just a matter or renaming the variable could help like  `Comment(ToHaveUpdatedBodyWithWholePluginsComment))`.

As for the `MockConfig` how about introducing key-value pairs just to avoid passing long strings with line separators.
```
mocker.MockConfig(ConfigYaml(
            WithParam("test_patterns", ['*my', 'test.go', 'pattern.js']").
            WithParam("skip_validation_for", "['pom.xml', 'regex{{*\\.adoc}}']").
            WithParam("plugin_hint", "http://my.server.com/message.md").
	ForChange(change)
```
Considering that the drawback and the overhead involved in this approach could be that we need to parse the whole `yaml`, may be could leave it as it is.

**Comment by MatousJobanek** [comment]
Thank you @hemanik for your ideas
> How about having LoadedFromDefaultJSON() and LoadedFromDefaultStruct() as a separate endpoint?

Good idea, if it makes the code more clear, then I can do that.

> Using the LoadedFrom() inside of the WithFiles() or ConfigYml() so we could directly pass the file path.

This is something that I had implemented in the very first version, but then I wanted to provide the content of the JSON directly in the method - eg. here: https://github.com/arquillian/ike-prow-plugins/pull/188/files#diff-0839f3fecac4bd84cfe03468fac9bfcfR114 - so this is the reason why it takes the content of the JSON by default and this is provided by the method `LoadedFrom()`.

> Also, I see that we are using the Expecting() part for checking matcher for the POST, DELETE or PATCH requests. Although it abstracts these details, but I somehow miss details about the change like if something is deleted or updated.
Rethinking on it with the DSL design may be just a matter or renaming the variable could help like Comment(ToHaveUpdatedBodyWithWholePluginsComment)).

The information if something is deleted or updated is visible only in the case of labels: https://github.com/arquillian/ike-prow-plugins/pull/188/files#diff-0839f3fecac4bd84cfe03468fac9bfcfR118 In the rest of the cases, there is no such information provided as all of them are only `POST` requests. Correct me if I'm wrong, the original version didn't contain such information either. Anyway, if it is missing there then I would stay with the proper naming of the variables.

> As for the MockConfig how about introducing key-value pairs just to avoid passing long strings with line separators.

Awesome idea! I really like it - I'll try to implement it there

File: pkg/command/cmd_executor_test.go [hunk 1]
```diff
@@ -167,34 +166,21 @@ var _ = Describe("Command executor features", func() {
...
+					Expecting(CommentTo(
+						HaveBodyThatContains("Hey @sender! It seems you tried to trigger `/command` command"),
+						HaveBodyThatContains("You have to be admin"))).
+					Create()
 
+				triggeredCommand.Repo = mock.PullRequest.Base.Repo
 				executed := false
 				command := is.CmdExecutor{Command: "/command"}
-				command.When(is.Triggered).By(user.Admin).Then(func() error {
+				command.When(is.Triggered).By(mock.CreateUserPermissionService("sender").Admin).Then(func() error {
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/command/cmd_executor_test.go]: you are right that it's hard to read.
I've change it to:
`mock.PermissionForUser("sender").ThatIs().Admin`
I hope that it is better

**Comment by MatousJobanek** [comment]
@hemanik @bartoszmajsak I've incorporated all your requested changes, so please review it when you have time.

**Pull request merged**